torch/csrc/api/src/data/samplers/distributed.cpp

```diff
@@ -0,0 +1,193 @@
+#include <torch/data/samplers/distributed.h>
+#include <torch/serialize/archive.h>
+#include <torch/types.h>
+
+#include <algorithm>
+#include <cstddef>
+#include <random>
+#include <vector>
+
+namespace torch {
+namespace data {
+namespace samplers {
+
+DistributedSampler::DistributedSampler(
+    size_t size,
+    size_t num_replicas,
+    size_t rank,
+    bool allow_duplicates)
+    : size_(size),
+      num_replicas_(num_replicas),
+      rank_(rank),
+      allow_duplicates_(allow_duplicates),
+      epoch_(0) {}
+
+void DistributedSampler::set_epoch(size_t epoch) {
+  epoch_ = epoch;
+}
+
+DistributedRandomSampler::DistributedRandomSampler(
+    size_t size,
+    size_t num_replicas,
+    size_t rank,
+    bool allow_duplicates)
+    : DistributedSampler(size, num_replicas, rank, allow_duplicates),
+      begin_index_(0),
+      end_index_(0),
+      sample_index_(0) {
+  populate_indices();
+  // shuffle first time.
+  reset(size_);
+}
+
+optional<std::vector<size_t>> DistributedRandomSampler::next(
+    size_t batch_size) {
+  if (sample_index_ == end_index_) {
+    return nullopt;
+  }
+
+  size_t end = sample_index_ + batch_size;
+  if (end > end_index_) {
+    end = end_index_;
+  }
+
+  auto iter = all_indices_.begin();
+  std::vector<size_t> res(iter + sample_index_, iter + end);
+  sample_index_ = end;
+  return res;
+}
+
+void DistributedRandomSampler::reset(optional<size_t> new_size) {
+  size_t size = new_size.value_or(size_);
+  if (size != size_) {
+    size_ = size;
+    populate_indices();
+  }
+  std::minstd_rand rand(epoch_);
+  std::shuffle(all_indices_.begin(), all_indices_.end(), rand);
+  sample_index_ = begin_index_;
+}
+
+void DistributedRandomSampler::populate_indices() {
+  size_t local_sample_count;
+  if (allow_duplicates_) {
+    local_sample_count = (size_ + num_replicas_ - 1) / num_replicas_;
+  } else {
+    local_sample_count = size_ / num_replicas_;
+  }
+  size_t sample_count =
+      num_replicas_ == 1 ? size_ : local_sample_count * num_replicas_;
+  all_indices_.resize(sample_count);
+  std::iota(std::begin(all_indices_), std::end(all_indices_), 0);
+  if (num_replicas_ > 1 && sample_count > size_) {
```

Comment:
Can we please eliminate this if? This check is completely redundant with the loop conditions, so it's pointless.

Reply:
Done.